### PR TITLE
Ongoing replication scalability suite

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

- adds minor updates, refactors and corrections to the brac data generation scripts
- switches pouchdb adapter to leveldb, to run 2-step scalability
- adds script that: 
  - generates initial data with configurable doc counts, and defaults: 
    - 5 district hospitals
    - 10 health centers per district hospital
    - 30 clinics per health center
    - 8 persons per clinic 
    - between 0-2 reports per person
    - 3 chws per health center
  - uploads initial data
  - runs jmeter test plan
- adds jmeter test plan that runs: 
  - initial replication (just like original scalability suite, except with a different adapter)
  - workflow replication
    - configurable iterations (default 10) and people (default 50)
    - opens existent local PouchDb database (the one created with the leveldb adapter during initial replication)
    - per iteration, generates configured number of persons and 0-2 reports per person (like in original data generation suite)
    - saves docs to local PouchDb
    - replicates up and down

ToDo: 
- finalize github workflow _that works!_ (right now it fails somewhere in the prep phase, where we're trying to upgrade the Amazon EC2 instance to use the _right_ version of the CHT). 

 
medic/cht-core#7077

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Review by garethbowen — COMMENTED** [review]
Couple of comments about naming. Didn't do a full review because it's still a draft. Seems like there's a lot of potential for reuse with the other scalability suite?

**Comment by dianabarsan** [comment]
> there's a lot of potential for reuse

Agree. 
Given I had no idea in which direction this was going, I opted for the two suites to be independent. 
I think we can factor in re-use when we've decided exactly how we're going to use this. Does that work for you?

File: .github/workflows/scalability-brac.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+name: BRAC Scalability Workflow
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply .github/workflows/scalability-brac.yml:1]: Cool. I agree.

**Comment by dianabarsan** [comment]
Thanks a lot for the review @garethbowen . 
Would you mind having a look at the PR description, which offers an overview of what I have right now, and share some thoughts? 

My main questions are: 
- do we want to run this in CI
- if yes, do we want to run this in CI in 3.x or just 4.x 
- if yes, do we just replace the previous scalability workflow with this one (since this also measures initial replication anyway). 
- what doc counts per user are we looking for, to set up correct doc counts.

**Comment by garethbowen** [comment]
> do we want to run this in CI

We want to be able to run it via GH Actions, but not in the main CI suite. I think eventually we'll have a nightly or weekly suite to run performance, scalability, etc testing but I think it's too slow and expensive to run on every commit, and with scalability there's no clear pass/fail anyway. For now either leave it as manual or fire it on beta just like the other scalability suite.

> do we want to run this in CI in 3.x or just 4.x

3.x would be useful to further prove the gains in 4.x but it's not a blocker if it's a lot of work for some reason.

> do we just replace the previous scalability workflow with this one (since this also measures initial replication anyway).

I think we need two numbers - (a) how many people can be onboarded in one go, and (b) how many can use the system at the same time after onboarding. I don't think it would be good to have one number which tried to measure both (a) and (b) because it won't be very accurate at either. I imagined this suite would do the initial replication _before_ the timer started so it would only test (b), leaving the other suite for (a). It would be possible (I think) to generate both numbers from a single jmeter run, in which case 1 suite is better than 2. But at this point go with whatever gives us (a) and (b) for the least amount of work.

> what doc counts per user are we looking for, to set up correct doc counts.

As close to production as possible, for a chosen production site. As the current ratios are based on BRAC can we roll with that? Is there a better instance to choose? Let's aim for close enough for now as every deployment will be different and it's easy to change the numbers later if we want to do a LG sized run for example.

**Comment by garethbowen** [comment]
@dianabarsan Where are we at with this PR?

**Comment by dianabarsan** [comment]
I kept using this PR to push code to the instances that ran the scalability script, so it became kind of messy. 
In terms of anything else, the last problem that I was facing was not being able to get the amazon instance up and running - not even the CHT one. 
If we rebase this to 4.0, installation will differ anyway, so we'll have to re-write that.

**Comment by dianabarsan** [comment]
Closing this PR as it's very stale, but keeping the branch.

**Pull request closed without merging**